File: web/src/app/topology/model/topology-contracts.ts
// TypeScript mirror of Caisson.Api/Contracts/TopologyContracts.cs and DiscoveryContracts.cs, field for
// field, in the camelCase the API's default System.Text.Json serialization produces on the wire. Keep
// this file in lock-step with the C# records — it is the only place the wire shape is declared.

export interface PagedResult<T> {
  items: T[];
  nextCursor: string | null;
}

export interface SnapshotMetadataDto {
  snapshotId: string;
  version: number;
  triggerType: string;
  createdBy: string;
  source: string;
  sourceVersion: string | null;
  createdAt: string;
  startedAt: string | null;
  completedAt: string | null;
  correlationId: string;
  status: string;
  diffSummary: unknown | null;
}

export interface SnapshotDetailDto {
  snapshot: SnapshotMetadataDto;
  graph: TopologyGraphDto;
}

export interface TopologyGraphDto {
  snapshotId: string;
  version: number;
  correlationId: string;
  servers: ServerNodeDto[];
  unmappedPorts: UnmappedPortDto[];
  /** Story #168: a flat switch → ports inventory, additive alongside the NIC-rooted graph above — the
   * Network Config Port Intent screen's source for "every discovered port on every switch". Optional so
   * every pre-existing graph fixture/test literal (written before this field existed) keeps compiling;
   * a real API response always includes it, so consumers should treat a missing value as `[]`. */
  switches?: SwitchInventoryDto[];
}

export interface ServerNodeDto {
  stableKey: string;
  hostname: string | null;
  bmcUuid: string | null;
  nics: NicNodeDto[];
}

/** `unmappedReasonCode` is set only when `bestAttachment` is null (backend story #10 step 1 addition). */
export interface NicNodeDto {
  stableKey: string;
  name: string;
  mac: string;
  bestAttachment: PortAttachmentDto | null;
  candidates: PortAttachmentDto[];
  unmappedReasonCode: string | null;
}

export interface PortAttachmentDto {
  switchStableKey: string;
  switchSerial: string | null;
  portName: string;
  confidence: number;
  band: string;
  reasonCode: string;
  vlans: number[];
}

export interface UnmappedPortDto {
  switchStableKey: string;
  switchSerial: string | null;
  portName: string;
}

/** A discovered switch and its full flat port inventory (story #168) — drives Network Config's Port
 * Intent screen. Existing NIC-rooted graph consumers ignore this field. */
export interface SwitchInventoryDto {
  stableKey: string;
  serial: string | null;
  name: string;
  ports: SwitchPortInventoryDto[];
}

export interface SwitchPortInventoryDto {
  stableKey: string;
  portName: string;
}

export interface EntityDiffDto {
  entityType: string;
  entityStableKey: string;
  changeType: string;
  payload: unknown;
  fromSnapshotId: string | null;
  toSnapshotId: string | null;
  createdAt: string | null;
  correlationId: string;
}

export interface SnapshotDiffDto {
  fromSnapshotId: string;
  toSnapshotId: string;
  changeSummary: unknown;
  diffs: EntityDiffDto[];
}

export interface EntityDetailDto {
  entityType: string;
  stableKey: string;
  latest: Record<string, string | null> | null;
  history: EntityDiffDto[];
}

export interface AuditEventDto {
  auditEventId: string;
  rackId: string | null;
  snapshotId: string | null;
  occurredAt: string;
  actorType: string;
  actorId: string;
  action: string;
  targetType: string;
  targetId: string | null;
  result: string;
  correlationId: string;
}

export interface DiscoveryJobSummaryDto {
  jobId: string;
  rackId: string;
  mode: string;
  status: string;
  createdAt: string;
  startedAt: string | null;
  finishedAt: string | null;
  triggeredBy: string;
  dryRun: boolean;
  errorCode: string | null;
  lastSuccessAt: string | null;
}

export interface DiscoveryStatusDto {
  rackId: string;
  latestJob: DiscoveryJobSummaryDto | null;
  lastSuccessAt: string | null;
  scheduleEnabled: boolean;
  nextRunAt: string | null;
}

File: src/Caisson.Api/Contracts/TopologyContracts.cs
using System.Text.Json;

namespace Caisson.Api.Contracts;

/// <summary>A single page of results with an opaque continuation cursor (null when exhausted).</summary>
public sealed record PagedResult<T>(IReadOnlyList<T> Items, string? NextCursor);

/// <summary>An observed rack available to the current principal.</summary>
public sealed record RackSummaryDto(Guid Id, string ExternalKey, string Name);

/// <summary>Snapshot metadata (AC1/AC3). <see cref="DiffSummary"/> is the change-count rollup, if any.</summary>
public sealed record SnapshotMetadataDto(
    Guid SnapshotId,
    int Version,
    string TriggerType,
    string CreatedBy,
    string Source,
    string? SourceVersion,
    DateTime CreatedAt,
    DateTime? StartedAt,
    DateTime? CompletedAt,
    Guid CorrelationId,
    string Status,
    JsonElement? DiffSummary);

/// <summary>A snapshot's metadata plus its projected topology graph (AC3 latest/detail).</summary>
public sealed record SnapshotDetailDto(SnapshotMetadataDto Snapshot, TopologyGraphDto Graph);

/// <summary>The projected topology graph for a snapshot (server-NIC → switch-port → VLAN).</summary>
public sealed record TopologyGraphDto(
    Guid SnapshotId,
    int Version,
    Guid CorrelationId,
    IReadOnlyList<ServerNodeDto> Servers,
    IReadOnlyList<UnmappedPortDto> UnmappedPorts,
    IReadOnlyList<SwitchInventoryDto> Switches);

/// <summary>A server and its NICs in the graph.</summary>
public sealed record ServerNodeDto(
    string StableKey,
    string? Hostname,
    string? BmcUuid,
    IReadOnlyList<NicNodeDto> Nics);

/// <summary>
/// A NIC with its best attachment and all candidate attachments. <see cref="UnmappedReasonCode"/> is
/// set only when the NIC has no attachment, explaining why (e.g. not seen in switch/BMC evidence).
/// </summary>
public sealed record NicNodeDto(
    string StableKey,
    string Name,
    string Mac,
    PortAttachmentDto? BestAttachment,
    IReadOnlyList<PortAttachmentDto> Candidates,
    string? UnmappedReasonCode);

/// <summary>A candidate NIC-to-port attachment with confidence, band, reason and VLANs.</summary>
public sealed record PortAttachmentDto(
    string SwitchStableKey,
    string? SwitchSerial,
    string PortName,
    double Confidence,
    string Band,
    string ReasonCode,
    IReadOnlyList<int> Vlans);

/// <summary>A switch port that no NIC mapped to.</summary>
public sealed record UnmappedPortDto(string SwitchStableKey, string? SwitchSerial, string PortName);

/// <summary>
/// A discovered switch and its full flat port inventory (story #168) — additive; drives the Network
/// Config Port Intent screen's switch/port selection. Existing NIC-rooted graph consumers ignore it.
/// </summary>
public sealed record SwitchInventoryDto(
    string StableKey, string? Serial, string Name, IReadOnlyList<SwitchPortInventoryDto> Ports);

/// <summary>A discovered port within a <see cref="SwitchInventoryDto"/>'s flat inventory.</summary>
public sealed record SwitchPortInventoryDto(string StableKey, string PortName);

/// <summary>A single per-entity diff (AC2 stored history or AC3 live drift).</summary>
public sealed record EntityDiffDto(
    string EntityType,
    string EntityStableKey,
    string ChangeType,
    JsonElement Payload,
    Guid? FromSnapshotId,
    Guid? ToSnapshotId,
    DateTime? CreatedAt,
    Guid CorrelationId);

/// <summary>The drift between two snapshots computed live (AC3 diff endpoint).</summary>
public sealed record SnapshotDiffDto(
    Guid FromSnapshotId,
    Guid ToSnapshotId,
    JsonElement ChangeSummary,
    IReadOnlyList<EntityDiffDto> Diffs);

/// <summary>An entity's current representation (if it still exists) plus its stored change history.</summary>
public sealed record EntityDetailDto(
    string EntityType,
    string StableKey,
    IReadOnlyDictionary<string, string?>? Latest,
    IReadOnlyList<EntityDiffDto> History);

/// <summary>An audit-trail event (AC3).</summary>
public sealed record AuditEventDto(
    Guid AuditEventId,
    Guid? RackId,
    Guid? SnapshotId,
    DateTime OccurredAt,
    string ActorType,
    string ActorId,
    string Action,
    string TargetType,
    string? TargetId,
    string Result,
    Guid CorrelationId);
